### PR TITLE
Add support for disabling use of TLS per domain suffix

### PR DESCRIPTION
**Pull request by cha7ri**
## Context:
We have clusters with two ingress controllers:
- Public ingress controller using the domain `example.com`
- Private ingress controller using the doamin `private.example.com`

## What we want to achieve?
- Use letsencrypt to issue certificates for `example.com`
- Disable tls for `private.example.com` 

## Solution:
- Set `use-ingress-tls` to `default_on` to enable tls by default for all ingresses
- Add a new config option to FIAAS to disable tls for one or many domains. example `tls-certificate-issuer-disable-for-domain-suffixes=private.example.com`

## What will change:
- Group hosts that we want to disable tls for together in one ingress resource (if the user doesn't set annotations).
- Do not add `kubernetes.io/tls-acme: true` or `cert-manager.io/cluster-issuer: nameOfClusterIssuer` annotation to the ingress resource.
- This will tell cert-manager to not create a certificate for those hosts

### CODE REVIEW

File: docs/operator_guide.md [hunk 1]
```diff
@@ -146,6 +146,7 @@ Used to configure how the ingress will be annotated for issuing a TLS certificat
 * `tls-certificate-issuer` sets the _value_ of the annotation, for example to decide between production and staging versions of an issuer in different namespaces
 * `tls-certificate-issuer-type-default` sets the default for the _key_ of the annotation, for example to use either `certmanager.k8s.io/cluster-issuer` (the default) or `certmanager.k8s.io/issuer`
 * `tls-certificate-issuer-type-overrides` allows specifying a mapping between the suffix of a domain and the issuer-type, to override the default. For example, assuming the 'cluster-issuer' type as the default, then specifying `--tls-certificate-issuer-type-overrides foo.example.com=certmanager.k8s.io/issuer` would mean that foo.example.com and any of its subdomains will use the 'issuer' type instead. In the case of multiple matching suffixes, the more specific (i.e. longest) will be used.
+* `disable-tls-for-domain-suffixes` disable tls for specified domain suffixes when `use-ingress-tls` is set to `default_on`. for example if you to enable tls for all domain suffixes except `internal.example.com` set `disable-tls-for-domain-suffixes internal.example.com` .
```
Comment (birgirst) [inline docs/operator_guide.md:149]: Consider adding a separate section around this perhaps above as this doesn't have the tls-certificate-issuer- prefix

Comment (cha7ri) [reply docs/operator_guide.md:151]: 👍

**Review by oyvindio — CHANGES_REQUESTED** [review]
I think the current implementation has behaviour that could be unexpected for some application configurations, see comment in-line.

File: tests/fiaas_deploy_daemon/test_config.py [hunk 2]
```diff
@@ -167,6 +167,14 @@ def test_tls_issuers(self):
             "foo.bar.com": "issuer",
             "woo.foo.bar.com": "other"
         }
+    def test_disable_tls_for_domain_suffixes(self):
+        disable_tls_for_domain_suffixes = ["foo.bar.com", "example.foo.bar.org"]
```
[the remark this answers is not in the document]
Comment (cha7ri) [reply tests/fiaas_deploy_daemon/test_config.py]: Fixed

File: fiaas_deploy_daemon/deployer/kubernetes/ingress.py [hunk 3]
```diff
@@ -279,11 +280,18 @@ def _collapse_hosts(self, app_spec, hosts):
...
 
-    def _should_have_ingress_tls(self, app_spec):
+    def _should_have_ingress_tls(self, app_spec, hosts):
         if self._use_ingress_tls == 'disabled' or app_spec.ingress_tls.enabled is False:
             return False
+        elif app_spec.ingress_tls.enabled is True:
+            return True
         else:
-            return self._use_ingress_tls == 'default_on' or app_spec.ingress_tls.enabled is True
+            #Check if any ingress host match or part of atleast one domain suffix that shoudln't have tls
```
[the remark this answers is not in the document]
Comment (cha7ri) [reply fiaas_deploy_daemon/deployer/kubernetes/ingress.py]: I fixed the group:

- Create an individual ingress for each annotated ingress
- Group no tls hosts together
- Group hosts that have tls issuer override together
- Group the rest in default ingress

I added unit tests to cover the use cases above and I added e2e tests

**Review by oyvindio — CHANGES_REQUESTED** [review]
Thanks for improving the tests and even adding e2e tests here. I've taken another look at this and left a few comments and questions in line

**Comment by arealmaas** [comment]
Any expected timeline for when this is to be merged? :)

**Comment by oyvindio** [comment]
I'm closing this as it is a long time since there was activity on this PR and the branch is stale. Feel free to re-open if it is still relevant.

**Pull request closed without merging**